File: apps/api/app/services/payment.py
"""Payment provider abstraction — Stripe test mode, local test provider fallback."""

from __future__ import annotations

import logging
from abc import ABC, abstractmethod

import httpx

from app.config import get_settings

logger = logging.getLogger(__name__)


class PaymentProvider(ABC):
    provider_name: str

    @abstractmethod
    async def create_payment_session(
        self,
        *,
        order_id: str,
        amount: int,
        currency: str,
        customer_email: str,
    ) -> dict:
        """Create a payment session and return provider metadata."""

    async def verify_webhook(self, payload: bytes, signature: str | None) -> dict | None:
        return None


class TestPaymentProvider(PaymentProvider):
    """Auto-succeeds so local/dev checkout can complete without Stripe keys."""

    provider_name = "test"

    async def create_payment_session(
        self,
        *,
        order_id: str,
        amount: int,
        currency: str,
        customer_email: str,
    ) -> dict:
        return {
            "status": "succeeded",
            "provider": self.provider_name,
            "orderId": order_id,
            "amount": amount,
            "currency": currency,
            "customerEmail": customer_email,
            "message": "Test payment captured. Add STRIPE_SECRET_KEY to use Stripe test mode.",
        }


class StripePaymentProvider(PaymentProvider):
    provider_name = "stripe"

    async def create_payment_session(
        self,
        *,
        order_id: str,
        amount: int,
        currency: str,
        customer_email: str,
    ) -> dict:
        settings = get_settings()
        stripe_amount = amount * 100 if currency.upper() == "INR" else amount
        async with httpx.AsyncClient(timeout=20.0) as client:
            response = await client.post(
                "https://api.stripe.com/v1/payment_intents",
                auth=(settings.stripe_secret_key, ""),
                data={
                    "amount": str(stripe_amount),
                    "currency": currency.lower(),
                    "receipt_email": customer_email,
                    "metadata[order_id]": order_id,
                    "automatic_payment_methods[enabled]": "true",
                },
            )
        if response.status_code >= 400:
            logger.warning("stripe.intent_failed %s", response.text)
            return {
                "status": "failed",
                "provider": self.provider_name,
                "orderId": order_id,
                "amount": amount,
                "currency": currency,
                "message": "Could not start Stripe checkout. Please try again.",
            }
        data = response.json()
        return {
            "status": "requires_action",
            "provider": self.provider_name,
            "orderId": order_id,
            "amount": amount,
            "currency": currency,
            "clientSecret": data.get("client_secret"),
            "paymentIntentId": data.get("id"),
            "publishableKey": settings.stripe_publishable_key,
            "message": "Complete payment with Stripe to confirm your order.",
        }


def get_payment_provider() -> PaymentProvider:
    settings = get_settings()
    if settings.stripe_secret_key.strip():
        return StripePaymentProvider()
    return TestPaymentProvider()

File: apps/web/lib/checkout/payment.ts
/**
 * Checkout payment session mapping.
 * Stripe test mode is created by the API when STRIPE_SECRET_KEY is set;
 * otherwise a local test provider confirms the order immediately.
 */

export type PaymentSession = {
  status: "coming_soon" | "ready" | "failed" | "succeeded" | "requires_action";
  provider: "stub" | "stripe" | "razorpay" | "test";
  orderId?: string;
  amount?: number;
  currency?: string;
  checkoutUrl?: string;
  clientSecret?: string;
  message?: string;
};

export interface PaymentProvider {
  initiateCheckout(params: {
    orderId: string;
    amount: number;
    currency: string;
    customerEmail: string;
  }): Promise<PaymentSession>;
}

class StubPaymentProvider implements PaymentProvider {
  async initiateCheckout(params: {
    orderId: string;
    amount: number;
    currency: string;
    customerEmail: string;
  }): Promise<PaymentSession> {
    return {
      status: "coming_soon",
      provider: "stub",
      orderId: params.orderId,
      amount: params.amount,
      currency: params.currency,
      message:
        "Payment integration pending — your order is saved. Secure checkout via Stripe or Razorpay will be enabled for launch.",
    };
  }
}

let provider: PaymentProvider = new StubPaymentProvider();

export function getPaymentProvider(): PaymentProvider {
  return provider;
}

/** Swap provider when Stripe/Razorpay is wired (e.g. from env). */
export function setPaymentProvider(next: PaymentProvider): void {
  provider = next;
}

export async function initiatePaymentFromCheckoutResponse(payment: {
  status: string;
  provider: string;
  orderId?: string;
  amount?: number;
  currency?: string;
  message?: string;
}): Promise<PaymentSession> {
  if (payment.provider === "stub" || payment.status === "coming_soon") {
    return getPaymentProvider().initiateCheckout({
      orderId: payment.orderId ?? "",
      amount: payment.amount ?? 0,
      currency: payment.currency ?? "INR",
      customerEmail: "",
    });
  }

  return {
    status: payment.status as PaymentSession["status"],
    provider: payment.provider as PaymentSession["provider"],
    orderId: payment.orderId,
    amount: payment.amount,
    currency: payment.currency,
    message: payment.message,
  };
}
